apollo-router/src/configuration/cors.rs

```diff
@@ -1,18 +1,173 @@
 use serde::Serialize;
-use tower_http::cors;
-use tower_http::cors::CorsLayer;
+use tower::Layer;
+use tower::Service;
+
+/// Configuration for a specific set of origins
+#[derive(Debug, Clone, Deserialize, Serialize, JsonSchema)]
+#[serde(deny_unknown_fields)]
+#[serde(default)]
+pub(crate) struct Policy {
+    /// Set to true to add the `Access-Control-Allow-Credentials` header for these origins
+    pub(crate) allow_credentials: Option<bool>,
+
+    /// The headers to allow for these origins
+    pub(crate) allow_headers: Vec<String>,
+
+    /// Which response headers should be made available to scripts running in the browser
+    pub(crate) expose_headers: Vec<String>,
+
+    /// `Regex`es you want to match the origins against to determine if they're allowed.
+    ///
+    /// - **Omit this field (do nothing):** Defaults to an empty list (no regex matching).
+    /// - `match_origins: []` (empty list): No regex matching.
+    ///
+    /// # Example
+    ///
+    /// ```yaml
+    /// # No regex matching (default)
+    /// policies:
+    ///   - origins: [https://myapp.com]
+    ///
+    /// # With regex matching
+    /// policies:
+    ///   - match_origins: ["^https://.*\\.example\\.com$"]
+    /// ```
+    #[serde(with = "serde_regex")]
+    #[schemars(with = "Vec<String>")]
+    pub(crate) match_origins: Vec<Regex>,
+
+    /// The `Access-Control-Max-Age` header value in time units
+    #[serde(deserialize_with = "humantime_serde::deserialize", default)]
+    #[schemars(with = "String", default)]
+    pub(crate) max_age: Option<Duration>,
+
+    /// Allowed request methods for these origins.
+    ///
+    /// **Note:** These defaults only apply when the entire `policies` field is omitted from the CORS config.
+    /// When specifying individual policies, this field defaults to an empty list unless explicitly set.
+    ///
+    /// - **When `policies` is omitted:** Defaults to `["GET", "POST", "OPTIONS"]` (uses default Policy).
+    /// - **Within a policy:** `methods: []` (empty list) means no methods allowed for this policy.
+    /// - **Within a policy:** `methods: [...]` uses the specified methods.
+    ///
+    /// # Examples
+    ///
+    /// ```yaml
+    /// # Use global default (Apollo Studio + default methods)
+    /// cors: {}
+    ///
+    /// # Within a policy - no methods allowed
+    /// cors:
+    ///   policies:
+    ///     - methods: []
+    ///
+    /// # Within a policy - custom methods
+    /// cors:
+    ///   policies:
+    ///     - methods: [GET, POST]
+    /// ```
+    pub(crate) methods: Vec<String>,
+
+    /// The origins to allow requests from.
+    ///
+    /// **Note:** These defaults only apply when the entire `policies` field is omitted from the CORS config.
+    /// When specifying individual policies, this field defaults to an empty list unless explicitly set.
+    ///
+    /// - **When `policies` is omitted:** Defaults to `["https://studio.apollographql.com"]` (uses default Policy).
```

Comment:
It seems confusing to document this case on a config object that you can only construct within a `policies` key. I think this documentation should exist only on the top-level cors configuration.